File: Editor/EasyRPG.Editor.Serializer/RM2K/LDB/LdbSkill.cs
using EasyRPG.Editor.RPG;

namespace EasyRPG.Editor.Serializer.RM2K.LDB
{
	public partial class LdbReader
	{
		#region BehaviorMethods
		private Skill ReadSkill()
		{
			var skill = new Skill { Id = stream.ReadCompressedInt() };

			while (!stream.EndOfFile())
			{
				var chunkId = (ChunkSkill)stream.ReadCompressedInt();
				if (chunkId == (int)ChunkData.End)
				{
					break;
				}

				var chunkLenght = stream.ReadCompressedInt();
				if (chunkLenght == 0)
				{
					continue;
				}

				switch (chunkId)
				{
					case ChunkSkill.Name:
						skill.Name = stream.ReadString(chunkLenght);
						break;
					case ChunkSkill.Description:
						skill.Description = stream.ReadString(chunkLenght);
						break;
					case ChunkSkill.UsingMessage1:
						skill.UsingMessage1 = stream.ReadString(chunkLenght);
						break;
					case ChunkSkill.UsingMessage2:
						skill.UsingMessage2 = stream.ReadString(chunkLenght);
						break;
					case ChunkSkill.FailureMessage:
						skill.FailureMessage = stream.ReadCompressedInt();
						break;
					case ChunkSkill.Type:
						skill.Type = (Skill.SkillType)stream.ReadCompressedInt();
						break;
					case ChunkSkill.SpType:
						skill.SpType = (Skill.SpTypeEnum)stream.ReadCompressedInt();
						break;
					case ChunkSkill.SpPercent:
						skill.SpPercent = stream.ReadCompressedInt();
						break;
					case ChunkSkill.SpCost:
						skill.SpCost = stream.ReadCompressedInt();
						break;
					case ChunkSkill.Scope:
						skill.Scope = (Skill.ScopeType)stream.ReadCompressedInt();
						break;
					case ChunkSkill.SwitchId:
						skill.SwitchId = stream.ReadCompressedInt();
						break;
					case ChunkSkill.AnimationId:
						skill.AnimationId = stream.ReadCompressedInt();
						break;
					case ChunkSkill.SoundEffect:
						skill.SoundEffect = ReadSound();
						break;
					case ChunkSkill.OccasionField:
						skill.OccasionField = stream.ReadCompressedBool();
						break;
					case ChunkSkill.OccasionBattle:
						skill.OccasionBattle = stream.ReadCompressedBool();
						break;
					case ChunkSkill.StateEffect:
						skill.StateEffect = stream.ReadCompressedBool();
						break;
					case ChunkSkill.PdefF:
						skill.PdefF = stream.ReadCompressedInt();
						break;
					case ChunkSkill.MdefF:
						skill.MdefF = stream.ReadCompressedInt();
						break;
					case ChunkSkill.Variance:
						skill.Variance = stream.ReadCompressedInt();
						break;
					case ChunkSkill.Power:
						skill.Power = stream.ReadCompressedInt();
						break;
					case ChunkSkill.Hit:
						skill.Hit = stream.ReadCompressedInt();
						break;
					case ChunkSkill.AffectHp:
						skill.AffectHp = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AffectSp:
						skill.AffectSp = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AffectAttack:
						skill.AffectAttack = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AffectDefense:
						skill.AffectDefense = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AffectSpirit:
						skill.AffectSpirit = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AffectAgility:
						skill.AffectAgility = stream.ReadCompressedBool();
						break;
					case ChunkSkill.AbsorbDamage:
						skill.AbsorbDamage = stream.ReadCompressedBool();
						break;
					case ChunkSkill.IgnoreDefense:
						skill.IgnoreDefense = stream.ReadCompressedBool();
						break;
					case ChunkSkill.StateSize:
						stream.ReadCompressedInt();
						break;
					case ChunkSkill.StateEffects:
						skill.StateEffects = stream.ReadBoolArray(chunkLenght);
						break;
					case ChunkSkill.AttributeSize:
						stream.ReadCompressedInt();
						break;
					case ChunkSkill.AttributeEffects:
						skill.AttributeEffects = stream.ReadBoolArray(chunkLenght);
						break;
					case ChunkSkill.AffectAttrDefence:
						skill.AffectAttrDefence = stream.ReadCompressedBool();
						break;
					case ChunkSkill.BattlerAnimation:
						skill.BattlerAnimation = stream.ReadCompressedInt();
						break;
					case ChunkSkill.BattlerAnimationData:
						for (var i = stream.ReadCompressedInt(); i > 0; i--)
						{
							skill.BattlerAnimationData.Add(ReadBattlerAnimationData());
						}
						break;
					default:
						stream.Skip(chunkLenght);
						break;
				}
			}
			return skill;
		}
		#endregion
	}
}

File: readers/include/detail/ldb_skill.cpp
/////////////////////////////////////////////////////////////////////////////
// This file is part of EasyRPG.
//
// EasyRPG is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// EasyRPG is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
/////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////
// Headers
////////////////////////////////////////////////////////////
#include "ldb_reader.h"
#include "ldb_chunks.h"
#include "reader_lcf.h"
#include "reader_struct.h"

////////////////////////////////////////////////////////////
/// Read Skill
////////////////////////////////////////////////////////////
EASYRPG_STRUCT_ID_READER(Skill, WithID)

EASYRPG_STRUCT_NAME(Skill)

#define EASYRPG_CHUNK_SUFFIX LDB_Reader
#define EASYRPG_CURRENT_STRUCT Skill

EASYRPG_STRUCT_FIELD_MAP_BEGIN(Skill)
	, EASYRPG_TYPED_FIELD_PAIR(std::string, name, ('name'))
	, EASYRPG_TYPED_FIELD_PAIR(std::string, description, ('desc')('ript')('ion'))
	, EASYRPG_TYPED_FIELD_PAIR(std::string, using_message1, ('usin')('g_me')('ssag')('e1'))
	, EASYRPG_TYPED_FIELD_PAIR(std::string, using_message2, ('usin')('g_me')('ssag')('e2'))
	, EASYRPG_TYPED_FIELD_PAIR(int, failure_message, ('fail')('ure_')('mess')('age'))
	, EASYRPG_TYPED_FIELD_PAIR(int, type, ('type'))
	, EASYRPG_TYPED_FIELD_PAIR(int, sp_type, ('sp_t')('ype'))
	, EASYRPG_TYPED_FIELD_PAIR(int, sp_percent, ('sp_p')('erce')('nt'))
	, EASYRPG_TYPED_FIELD_PAIR(int, sp_cost, ('sp_c')('ost'))
	, EASYRPG_TYPED_FIELD_PAIR(int, scope, ('scop')('e'))
	, EASYRPG_TYPED_FIELD_PAIR(int, switch_id, ('swit')('ch_i')('d'))
	, EASYRPG_TYPED_FIELD_PAIR(int, animation_id, ('anim')('atio')('n_id'))
	, EASYRPG_TYPED_FIELD_PAIR(RPG::Sound, sound_effect, ('soun')('d_ef')('fect'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, occasion_field, ('occa')('sion')('_fie')('ld'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, occasion_battle, ('occa')('sion')('_bat')('tle'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, state_effect, ('stat')('e_ef')('fect'))
	, EASYRPG_TYPED_FIELD_PAIR(int, pdef_f, ('pdef')('_f'))
	, EASYRPG_TYPED_FIELD_PAIR(int, mdef_f, ('mdef')('_f'))
	, EASYRPG_TYPED_FIELD_PAIR(int, variance, ('vari')('ance'))
	, EASYRPG_TYPED_FIELD_PAIR(int, power, ('powe')('r'))
	, EASYRPG_TYPED_FIELD_PAIR(int, hit, ('hit'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_hp, ('affe')('ct_h')('p'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_sp, ('affe')('ct_s')('p'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_attack, ('affe')('ct_a')('ttac')('k'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_defense, ('affe')('ct_d')('efen')('se'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_spirit, ('affe')('ct_s')('piri')('t'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_agility, ('affe')('ct_a')('gili')('ty'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, absorb_damage, ('abso')('rb_d')('amag')('e'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, ignore_defense, ('igno')('re_d')('efen')('se'))
	, EASYRPG_SIZE_FIELD_PAIR(bool, state_effects, ('stat')('e_ef')('fect')('s'))
	, EASYRPG_TYPED_FIELD_PAIR(std::vector<bool>, state_effects, ('stat')('e_ef')('fect')('s'))
	, EASYRPG_SIZE_FIELD_PAIR(bool, attribute_effects, ('attr')('ibut')('e_ef')('fect')('s'))
	, EASYRPG_TYPED_FIELD_PAIR(std::vector<bool>, attribute_effects, ('attr')('ibut')('e_ef')('fect')('s'))
	, EASYRPG_TYPED_FIELD_PAIR(bool, affect_attr_defence, ('affe')('ct_a')('ttr_')('defe')('nce'))
	, EASYRPG_TYPED_FIELD_PAIR(int, battler_animation, ('batt')('ler_')('anim')('atio')('n'))
	, EASYRPG_TYPED_FIELD_PAIR(std::vector<RPG::BattlerAnimationData>, battler_animation_data, ('batt')('ler_')('anim')('atio')('n_da')('ta'))
EASYRPG_STRUCT_FIELD_MAP_END()

#undef EASYRPG_CURRENT_STRUCT
#undef EASYRPG_CHUNK_SUFFIX
